README
DBD::TimesTen -- DBD module interfacing the TimesTen databases.

   $Id: README 518 2006-11-22 22:51:21Z wagnerch $

   Copyright (c) 1994,1995,1996,1997,1998  Tim Bunce
   portions Copyright (c) 1997,1998,1999,2000,2001,2002 Jeff Urlwin
   portions Copyright (c) 1997  Thomas K. Wenrich
   portions Copyright (c) 2006  Chad Wagner

   You may distribute under the terms of either the GNU General Public
   License or the Artistic License, as specified in the Perl README file.

   PLEASE READ THE ENTIRE README FILE CAREFULLY !


**** NOTICE ******** NOTICE ******** NOTICE ******** NOTICE ******** NOTICE ****

   This is a BETA version of the driver, since I am not actively using TimesTen
   I have not encountered every issue.  This driver is based on DBD::ODBC,
   since the TimesTen client can be directly linked to any ODBC-based
   application.  The intent is to eventually code all of the TimesTen
   extensions into this driver.

**** NOTICE ******** NOTICE ******** NOTICE ******** NOTICE ******** NOTICE ****


INTRODUCTION:
   This driver is based on DBD::ODBC 1.13.  Thanks to Tim Bunce, Jeff Urlwin,
   and the contributors of DBD::ODBC, as they have made this driver fairly
   easy to implement with TimesTen.

   There is more work to be done here, it appears the bulk inserts need to
   be done and probably a bunch of other things.


*BEFORE* BUILDING, TESTING AND INSTALLING this you will need to:

    Build, test and install Perl 5 (as per DBI specifications/compatibility)
    It is very important to TEST it and INSTALL it!

    Build, test and install the DBI module (at least DBI 1.201).
    It is very important to TEST it and INSTALL it!

    Remember to *read* the DBI README file!

    Remember to *read* the DBD::TimesTen.pm POD documentation!


BUILDING:

  set-up these environment variables:
    DBI_DSN   The dbi data source, e.g. 'dbi:TimesTen:YOUR_DSN_HERE'
    DBI_USER  The username to use to connect to the database
    DBI_PASS  The username to use to connect to the database
    TT_HOME   (Unix only) The dir your driver manager is installed in
  perl Makefile.PL
  make                (or nmake, if VC++ on Win32)
  make test           (or nmake, if VC++ on Win32)

TESTING
  
  make test
  make test TEST_VERBOSE=1   (if any of the t/* tests fail)
  make install               (if the tests look okay)


IF YOU HAVE PROBLEMS:

Do not hand edit the generated Makefile unless you are completely sure
you understand the implications! Always try to make changes via the
Makefile.PL command line and/or editing the Makefile.PL.

You should not need to make any changes. If you do *please* let me
know so that I can try to make it automatic in a later release.

Please post details of any problems (or changes you needed to make) to
chad.wagner@gmail.com. But note...

** IT IS IMPORTANT TO INCLUDE THE FOLLOWING INFORMATION:

1. A complete log of a all steps of the build, e.g.:

  perl Makefile.PL           (do a make realclean first)
  make
  make test
  make test TEST_VERBOSE=1   (if any tests fail)

2. Full details of which software you are using, including:

  Perl version (the output of perl -V)

It is important to check that you are using the latest version before
posting.


===============================================================================
Other info:

DBI 'home page': http://dbi.perl.org

perldoc DBI
perldoc DBD::TimesTen

End.


Changes
=head1 NAME

DBD::TimesTen::Changes - Log of significant changes to the DBD::TimesTen

=cut

=head1 CHANGES

=head2 Changes in DBD::TimesTen 0.02 November 22, 2006

Changed prefix of functions from odbc to timesten, to avoid namespace
  conflicts.

=head2 Changes in DBD::TimesTen 0.01 November 22, 2006

Forked code base from DBD::ODBC 1.13

=cut


t/ODBCTEST.pm
#
# Package ODBCTEST
# 
# This package is a common set of routines for the DBD::ODBC tests.
# This is a set of routines to create, drop and test for existance of
# a table for a given DBI database handle (dbh).
#
# This set of routines currently depends greatly upon some ODBC meta-data.
# The meta data required is the driver's native type name for various ODBC/DBI
# SQL types.  For example, SQL_VARCHAR would produce VARCHAR2 under Oracle and TEXT
# under MS-Access.  This uses the function SQLGetTypeInfo.  This is obtained via
# the DBI C<func> method, which is implemented as a call to the driver.  In this case,
# of course, this is the DBD::ODBC.
#
# the SQL_TIMESTAMP may be dubious on many platforms, but SQL_DATE was not supported
# under Oracle, MS SQL Server or Access.  Those are pretty common ones.
#

require 5.004;
{
   package ODBCTEST;

   use DBI qw(:sql_types);
   use Test::More;
   
   $VERSION = '0.01';
   $table_name = "PERL_DBD_TEST";

   $longstr = "THIS IS A STRING LONGER THAN 80 CHARS.  THIS SHOULD BE CHECKED FOR TRUNCATION AND COMPARED WITH ITSELF.";
   $longstr2 = $longstr . "  " . $longstr . "  " . $longstr . "  " . $longstr;

    # really dumb work around:
    # MS SQL Server 2000 (MDAC 2.5 and ODBC driver 2000.080.0194.00) have a bug if
    # the column is named C, CA, or CAS and there is a call to SQLDescribeParam...
    # there is an error, referring to a syntax error near keyword 'by'
    # I figured it's just best to rename the columns.
    # changed SQL_BIGINT below to -5, as DBI has removed that constant.
    # ODBC's value is -5.
   %TestFieldInfo = (
		     'COL_A' => [SQL_SMALLINT,-5, SQL_TINYINT, SQL_NUMERIC, SQL_DECIMAL, SQL_FLOAT, SQL_REAL],
		     'COL_B' => [SQL_VARCHAR, SQL_CHAR],
		     'COL_C' => [SQL_VARCHAR],
		     'COL_D' => [SQL_TYPE_TIMESTAMP, SQL_TYPE_DATE, SQL_DATE, SQL_TIMESTAMP ],
		    );

   sub get_type_for_column {
      my $dbh = shift;
      my $column = shift;

      my $type;
      my @row;
      my $sth;
      foreach $type (@{ $TestFieldInfo{$column} }) {
	 $sth = $dbh->func($type, GetTypeInfo);
	    # may not be correct behavior, but get the first compat type
	 if ($sth) {
	    @row = $sth->fetchrow();
	    $sth->finish();
	    last if @row;
	 } else {
		    # warn "Unable to get type for type $type\n";
	 }
      }
      die "Unable to find a suitable test type for field $column"
	    unless @row;
	# warn join(", ",@row);
      return @row;
   }
   sub tab_create {
      my $dbh = shift;
      $dbh->{PrintError} = 0;
      eval {
	 $dbh->do("DROP TABLE $table_name");
      };
      $dbh->{PrintError} = 1;

      # trying to use ODBC to tell us what type of data to use
      my $fields = undef;
      my $f;
      foreach $f (sort keys %TestFieldInfo) {
	    # print "$f: $TestFieldInfo{$f}\n";
	 $fields .= ", " unless !$fields;
	 $fields .= "$f ";
	    # print "-- $fields\n";
	 my @row = get_type_for_column($dbh, $f);
	 $fields .= $row[0];
	 if ($row[5]) {
	    $fields .= "($row[2])"	 if ($row[5] =~ /LENGTH/i);
	    $fields .= "($row[2],0)" if ($row[5] =~ /PRECISION,SCALE/i);
	 }
	 if ($f eq 'COL_A') {
	    $fields .= " NOT NULL PRIMARY KEY ";
	 }
	    # print "-- $fields\n";
      }
      # diag("Using fields: $fields\n");
      $dbh->do("CREATE TABLE $table_name ($fields)");
   }


   sub tab_delete {
      my $dbh = shift;
      $dbh->do("DELETE FROM $table_name");
   }

   sub tab_exists {
      my $dbh = shift;
      my (@rows, @row, $rc);

      $rc = -1;

      unless ($sth = $dbh->table_info()) {
	 # diag("Can't list tables: $DBI::errstr\n");
	 return -1;
      }
	# TABLE_QUALIFIER,TABLE_OWNER,TABLE_NAME,TABLE_TYPE,REMARKS
      while ($row = $sth->fetchrow_hashref()) {
	    # XXX not fully true.  The "owner" could be different.  Need to check!
	    # In Oracle, testing $user against $row[1] works, but does NOT in SQL Server.
	    # SQL server returns the device and something else I haven't quite taken the time
	    # to figure it out, since I'm not a SQL server expert.  Anyone out there?
	    # (mine returns "dbo" for the owner on ALL my tables.  This is obviously something
	    # significant for SQL Server...one of these days I'll dig...
	 if (($table_name eq uc($row->{TABLE_NAME}))) {
		# and (uc($user) eq uc($row[1]))) 
		# qeDBF driver returns null for TABLE_OWNER
	    my $owner = $row->{TABLE_OWNER} || '(unknown owner)';
	    # diag("$owner.$row->{TABLE_NAME}\n");
	    $rc = 1;
	    last;
	 }
      }
      $sth->finish();
      $rc;
   }

   #
   # show various ways of inserting data without binding parameters.
   # Note, these are not necessarily GOOD ways to
   # show this...
   #

   @tab_insert_values = ( 
			 [1, 'foo', 'foo varchar', "{d '1998-05-11'}", "{ts '1998-05-11 00:00:00'}"],
			 [2, 'bar', 'bar varchar', "{d '1998-05-12'}", "{ts '1998-05-12 00:00:00'}"],
			 [3, "bletch", "bletch varchar", "{d '1998-05-10'}", "{ts '1998-05-10 00:00:00'}"],
			 [4, "80char", $longstr, "{d '1998-05-13'}", "{ts '1998-05-13 12:00:00'}"],
			 [5, "gt250char", $longstr2, "{d '1998-05-14'}", "{ts '1998-05-14 00:00:00'}"],
			);
			   
   sub tab_insert {
      my $dbh = shift;

       # qeDBF needs a space after the table name!
      foreach (@tab_insert_values) {
      
	 @row = ODBCTEST::get_type_for_column($dbh, 'COL_D');
	 # print "TYPE FOUND = $row[1]\n";
	 if (!$dbh->do("INSERT INTO $table_name (COL_A, COL_B, COL_C, COL_D) VALUES ("
		 . join(", ", $_->[0],
			$dbh->quote($_->[1]),
			$dbh->quote($_->[2]), 
			$_->[isDateType($row[1]) ? 3 : 4]). ")")) {
	    return 0;
	 }
      }
      1;
   }

   sub isDateType($) {
      my $type = shift;
      if ($type == SQL_DATE  || $type == SQL_TYPE_DATE) {
	 return 1;
      } else {
	 return 0;
      }
   }

   sub tab_insert_bind {
      my $dbh = shift;
      my $dref = shift;
      my $handle_column_type = shift;
      my @data = @{$dref};

      my $sth = $dbh->prepare("INSERT INTO $table_name (COL_A, COL_B, COL_C, COL_D) VALUES (?, ?, ?, ?)");
      unless ($sth) {
	 warn $DBI::errstr;
	 return 0;
      }
      # $sth->{PrintError} = 1;
      foreach (@data) {
	 my @row;
	 if ($handle_column_type) {
	    @row = ODBCTEST::get_type_for_column($dbh, 'COL_A');
	    # diag("Binding the value: $_->[0] type = $row[1]\n");
	    $sth->bind_param(1, $_->[0], { TYPE => $row[1] });
	 } else {
	    $sth->bind_param(1, $_->[0]);
	 }
	 if ($handle_column_type) {
	    @row = ODBCTEST::get_type_for_column($dbh, 'COL_B');
	    $sth->bind_param(2, $_->[1], { TYPE => $row[1] });
	 } else {
	    $sth->bind_param(2, $_->[1]);
	 }
	 if ($handle_column_type) {
	    @row = ODBCTEST::get_type_for_column($dbh, 'COL_C');
	    $sth->bind_param(3, $_->[2], { TYPE => $row[1] });
	 } else {
	    $sth->bind_param(3, $_->[2]);
	 }
	 
	 # print "SQL_DATE = ", SQL_DATE, " SQL_TIMESTAMP = ", SQL_TIMESTAMP, "\n";
	 @row = ODBCTEST::get_type_for_column($dbh, 'COL_D');
	 # diag("TYPE FOUND = $row[1]\n");
	 # if ($row[1] == SQL_TYPE_TIMESTAMP) {
	 #   $row[1] = SQL_TIMESTAMP;
	 #}
	 # print "Binding the date value: \"$_->[$row[1] == SQL_DATE ? 3 : 4]\"\n";
	 if ($handle_column_type) {
	    $sth->bind_param(4, $_->[isDateType($row[1]) ? 3 : 4], { TYPE => $row[1] });
	 } else {
	    # sigh, couldn't figure out how to get rid of the warning nicely,
	    # so I turned it off!!!  Now, I have to turn it back on due
	    # to  problems in other perl versions.
	    $sth->bind_param(4, $_->[isDateType($row[1]) ? 3 : 4]);
	 }
	 return 0 unless $sth->execute;
      }
      1;
   }
   1;
}



TimesTen.h
/*
 * $Id: TimesTen.h 530 2006-11-25 21:39:51Z wagnerch $
 * Copyright (c) 1994,1995,1996,1997  Tim Bunce
 *
 * You may distribute under the terms of either the GNU General Public
 * License or the Artistic License, as specified in the Perl README file.
 */

#include <timesten.h>

#define NEED_DBIXS_VERSION 9

#include <DBIXS.h>	/* from DBI. Load this after dbdodbc.h */

#include "dbivport.h" /* copied from DBI to maintain compatibility */

#include "dbdimp.h"

#include <dbd_xsh.h>	/* from DBI. Load this after dbdodbc.h */

SV      *dbd_db_get_info _((SV *dbh, int ftype));
int      dbd_db_type_info _((SV *dbh, SV *sth, int ftype));
SV	*dbd_st_cancel _((SV *sth));
int	 dbd_db_column_info _((SV *dbh, SV *sth,
	    char *catalog, char *schema, char *table, char *column));

int  dbd_db_table_info _((SV *dbh, SV *sth, char *catalog, char *schema, char *table, char *table_type));
int  dbd_db_primary_key_info _((SV *dbh, SV *sth, char *catalog, char *schema, char *table));
int  dbd_db_foreign_key_info _((SV *dbh, SV *sth, char *PK_CatalogName, char *PK_SchemaName, char *PK_TableName, char *FK_CatalogName, char *FK_SchemaName, char *FK_TableName));
void dbd_error _((SV *h, RETCODE err_rc, char *what));
void dbd_error2 _((SV *h, RETCODE err_rc, char *what, HENV henv, HDBC hdbc, HSTMT hstmt));
int dbd_db_execdirect _(( SV *dbh, char *statement ));


/* end of ODBC.h */


Makefile.PL
#!/usr/local/bin/perl -w
#
# $Id: Makefile.PL 526 2006-11-25 17:34:51Z wagnerch $
#
# You may distribute under the terms of either the GNU General Public
# License or the Artistic License, as specified in the Perl README file.
#
BEGIN { require 5.004 }	# 5.004 is required for Win32
use Config;
use ExtUtils::MakeMaker 5.16, qw(&WriteMakefile $Verbose);
use File::Basename;
use Getopt::Long;
use File::Spec;

use DBI 1.21;		# must be installed first ...
use DBI::DBD;
use strict;

my %opts = (
   NAME	=> 'DBD::TimesTen',
   VERSION_FROM => 'TimesTen.pm',
   PREREQ_PM => { "Test::Simple" => 0.40 }, # actually Test::More pkg in T::S dist
   clean => { FILES=> 'TimesTen.xsi' },
   dist => {
      DIST_DEFAULT=> 'clean svnlog distcheck tardist',
      PREOP    => '$(MAKE) -f Makefile.old distdir',
      COMPRESS => 'gzip -v9', SUFFIX => 'gz'
   },
   OBJECT => '$(O_FILES)',
   DEFINE => '',
);

if ($ExtUtils::MakeMaker::VERSION >= 5.43) {
   $opts{AUTHOR} = 'Tim Bunce and Jeff Urlwin mailto:dbi-users@perl.org';
   $opts{ABSTRACT} = 'TimesTen driver for the DBI module.';
   $opts{PREREQ_PM} = { DBI => 1.21 };
   $opts{CAPI} = 'TRUE' if $Config{archname} =~ /-object\b/i;
}

our $opt_g = 0;
our $opt_o = '';
Getopt::Long::GetOptions (
    "g!" => \$opt_g
   ,"o=s" => \$opt_o
) or die "Invalid arguments";

print "Overriding TimesTen Directory with command line option: $opt_o\n"
   if $opt_o;

if (-d ".svn")
{
   $opts{DEFINE} .= ' -Wall';
   $opt_g = 1;
}

if ($opt_g)
{
   print "Setting debug options!\n";
   $opts{OPTIMIZE} = '-g -O0';
}

print "\nConfiguring DBD::TimesTen ...\n
>>>\tRemember to actually *READ* the README file!
   \tAnd re-read it if you have any problems.\n
";

my $dbi_dir      = dbd_dbi_dir();
my $dbi_arch_dir = dbd_dbi_arch_dir();
my $timesten_home= $opt_o || $ENV{TT_HOME};

unless ($timesten_home)
{
   print qq{
The DBD::TimesTen module needs to link with the TimesTen client library.

You need to indicate where your TimesTen client is installed.
You can do this ether by setting the TT_HOME environment variable
or by runing 'perl Makefile.PL -o <ttdir>'.

};

   die "Makefile.PL aborted.\n";
}

die "TT_HOME environment variable ($timesten_home) does not refer to a directory.\n"
   unless -d $timesten_home;

warn "Using TimesTen in $timesten_home\n";

$opts{INC} = "-I. -I$dbi_arch_dir -I$timesten_home/include";
$opts{LIBS} = "-L$timesten_home/lib -lttclient";
$opts{LDDLFLAGS} = $Config{lddlflags} . " -Wl,-rpath,$timesten_home/lib";

print "\n";
my $rv = WriteMakefile(%opts);

local($^W)=0;
print qq{
The DBD::TimesTen tests will use these values for the database connection:

    DBI_DSN=$ENV{DBI_DSN}		e.g. dbi:TimesTen:demo
    DBI_USER=$ENV{DBI_USER}
    DBI_PASS=$ENV{DBI_PASS}
};
print "Warning: not all required environment variables are set.\n"
   unless ($ENV{DBI_DSN} && $ENV{DBI_USER} && $ENV{DBI_PASS});

print "Warning: DBI_DSN ($ENV{DBI_DSN}) doesn't start with 'dbi:TimesTen:'\n"
   if ($ENV{DBI_DSN} && $ENV{DBI_DSN} !~ m/^dbi:TimesTen:/i);

print "\n";

$rv;

# ====================================================================

{
   package MY; # SUPER needs package context, $self is not sufficient

   use strict;
   use Config;

   sub postamble
   {
      return main::dbd_postamble(@_);
   }

   sub dist_basics
   {
      return shift->SUPER::dist_basics(@_) . '
svnlog :
	svn log -v --xml svn://localhost/repos/dbd-timesten/ |$(SVN2LOG)/svn2log.py -D 0 -s -O -L -H -F -u "$(SVN2LOG)/svn2log.users" -x "dbd-timesten/branches/"
';
   }
}
__END__




TimesTen.pm
# $Id: TimesTen.pm 544 2006-11-26 15:47:54Z wagnerch $
#
# Copyright (c) 1994,1995,1996,1998  Tim Bunce
# portions Copyright (c) 1997-2004  Jeff Urlwin
# portions Copyright (c) 1997  Thomas K. Wenrich
# portions Copyright (c) 2006  Chad Wagner
#
# You may distribute under the terms of either the GNU General Public
# License or the Artistic License, as specified in the Perl README file.

require 5.004;

$DBD::TimesTen::VERSION = '0.04';

{
    package DBD::TimesTen;

    use DBI ();
    use DynaLoader ();
    use Exporter ();
    
    @ISA = qw(Exporter DynaLoader);
    %EXPORT_TAGS = (
       sql_isolation_options => [ qw(
          SQL_TXN_READ_COMMITTED SQL_TXN_SERIALIZABLE
       ) ],
       sql_getinfo_options => [ qw(
          SQL_INFO_FIRST SQL_ACTIVE_CONNECTIONS SQL_ACTIVE_STATEMENTS
          SQL_DATA_SOURCE_NAME SQL_DRIVER_HDBC SQL_DRIVER_HENV SQL_DRIVER_HSTMT
          SQL_DRIVER_NAME SQL_DRIVER_VER SQL_FETCH_DIRECTION
          SQL_ODBC_API_CONFORMANCE SQL_ODBC_VER SQL_ROW_UPDATES
          SQL_ODBC_SAG_CLI_CONFORMANCE SQL_SERVER_NAME
          SQL_SEARCH_PATTERN_ESCAPE SQL_ODBC_SQL_CONFORMANCE SQL_DBMS_NAME
          SQL_DBMS_VER SQL_ACCESSIBLE_TABLES SQL_ACCESSIBLE_PROCEDURES
          SQL_PROCEDURES SQL_CONCAT_NULL_BEHAVIOR SQL_CURSOR_COMMIT_BEHAVIOR
          SQL_CURSOR_ROLLBACK_BEHAVIOR SQL_DATA_SOURCE_READ_ONLY
          SQL_DEFAULT_TXN_ISOLATION SQL_EXPRESSIONS_IN_ORDERBY
          SQL_IDENTIFIER_CASE SQL_IDENTIFIER_QUOTE_CHAR
          SQL_MAX_COLUMN_NAME_LEN SQL_MAX_CURSOR_NAME_LEN
          SQL_MAX_OWNER_NAME_LEN SQL_MAX_PROCEDURE_NAME_LEN
          SQL_MAX_QUALIFIER_NAME_LEN SQL_MAX_TABLE_NAME_LEN
          SQL_MULT_RESULT_SETS SQL_MULTIPLE_ACTIVE_TXN SQL_OUTER_JOINS
          SQL_OWNER_TERM SQL_PROCEDURE_TERM SQL_QUALIFIER_NAME_SEPARATOR
          SQL_QUALIFIER_TERM SQL_SCROLL_CONCURRENCY SQL_SCROLL_OPTIONS
          SQL_TABLE_TERM SQL_TXN_CAPABLE SQL_USER_NAME SQL_CONVERT_FUNCTIONS
          SQL_NUMERIC_FUNCTIONS SQL_STRING_FUNCTIONS SQL_SYSTEM_FUNCTIONS
          SQL_TIMEDATE_FUNCTIONS SQL_CONVERT_BIGINT SQL_CONVERT_BINARY
          SQL_CONVERT_BIT SQL_CONVERT_CHAR SQL_CONVERT_DATE SQL_CONVERT_DECIMAL
          SQL_CONVERT_DOUBLE SQL_CONVERT_FLOAT SQL_CONVERT_INTEGER
          SQL_CONVERT_LONGVARCHAR SQL_CONVERT_NUMERIC SQL_CONVERT_REAL
          SQL_CONVERT_SMALLINT SQL_CONVERT_TIME SQL_CONVERT_TIMESTAMP
          SQL_CONVERT_TINYINT SQL_CONVERT_VARBINARY SQL_CONVERT_VARCHAR
          SQL_CONVERT_LONGVARBINARY SQL_TXN_ISOLATION_OPTION
          SQL_ODBC_SQL_OPT_IEF SQL_CORRELATION_NAME SQL_NON_NULLABLE_COLUMNS
          SQL_DRIVER_HLIB SQL_DRIVER_ODBC_VER SQL_LOCK_TYPES SQL_POS_OPERATIONS
          SQL_POSITIONED_STATEMENTS SQL_GETDATA_EXTENSIONS
          SQL_BOOKMARK_PERSISTENCE SQL_STATIC_SENSITIVITY SQL_FILE_USAGE
          SQL_NULL_COLLATION SQL_ALTER_TABLE SQL_COLUMN_ALIAS SQL_GROUP_BY
          SQL_KEYWORDS SQL_ORDER_BY_COLUMNS_IN_SELECT SQL_OWNER_USAGE
          SQL_QUALIFIER_USAGE SQL_QUOTED_IDENTIFIER_CASE SQL_SPECIAL_CHARACTERS
          SQL_SUBQUERIES SQL_UNION SQL_MAX_COLUMNS_IN_GROUP_BY
          SQL_MAX_COLUMNS_IN_INDEX SQL_MAX_COLUMNS_IN_ORDER_BY
          SQL_MAX_COLUMNS_IN_SELECT SQL_MAX_COLUMNS_IN_TABLE SQL_MAX_INDEX_SIZE
          SQL_MAX_ROW_SIZE_INCLUDES_LONG SQL_MAX_ROW_SIZE SQL_MAX_STATEMENT_LEN
          SQL_MAX_TABLES_IN_SELECT SQL_MAX_USER_NAME_LEN
          SQL_MAX_CHAR_LITERAL_LEN SQL_TIMEDATE_ADD_INTERVALS
          SQL_TIMEDATE_DIFF_INTERVALS SQL_NEED_LONG_DATA_LEN
          SQL_MAX_BINARY_LITERAL_LEN SQL_LIKE_ESCAPE_CLAUSE
          SQL_QUALIFIER_LOCATION SQL_INFO_LAST
       ) ],
    );
    Exporter::export_ok_tags(qw(sql_isolation_options sql_getinfo_options));

    require_version DBI 1.21;

    bootstrap DBD::TimesTen $VERSION;

    $err = 0;		# holds error code   for DBI::err
    $errstr = "";	# holds error string for DBI::errstr
    $sqlstate = "00000";
    $drh = undef;	# holds driver handle once initialised

    sub driver{
	return $drh if $drh;
	my($class, $attr) = @_;

	$class .= "::dr";

	# not a 'my' since we use it above to prevent multiple drivers

	$drh = DBI::_new_drh($class, {
	    'Name' => 'TimesTen',
	    'Version' => $VERSION,
	    'Err'    => \$DBD::TimesTen::err,
	    'Errstr' => \$DBD::TimesTen::errstr,
	    'State' => \$DBD::TimesTen::sqlstate,
	    'Attribution' => 'TimesTen DBD by Chad Wagner',
	    });

	$drh;
    }

    sub CLONE { undef $drh }

    sub AUTOLOAD {
        (my $constname = $AUTOLOAD) =~ s/.*:://;
        my $val = constant($constname); 
        *$AUTOLOAD = sub { $val };
        goto &$AUTOLOAD;
    }

    1;
}


{   package DBD::TimesTen::dr; # ====== DRIVER ======
    use strict;

    sub connect {
	my $drh = shift;
	my($dbname, $user, $auth, $attr)= @_;
	$user = '' unless defined $user;
	$auth = '' unless defined $auth;

	# create a 'blank' dbh
	my $this = DBI::_new_dbh($drh, {
	    'Name' => $dbname,
	    'USER' => $user, 
	    'CURRENT_USER' => $user,
	    });

	# Call ODBC logon func in ODBC.xs file
	# and populate internal handle data.

	DBD::TimesTen::db::_login($this, $dbname, $user, $auth, $attr) or return undef;

	$this;
    }

}


{   package DBD::TimesTen::db; # ====== DATABASE ======
    use strict;

    sub prepare {
	my($dbh, $statement, @attribs)= @_;

	# create a 'blank' dbh
	my $sth = DBI::_new_sth($dbh, {
	    'Statement' => $statement,
	    });

	# Call ODBC func in ODBC.xs file.
	# (This will actually also call SQLPrepare for you.)
	# and populate internal handle data.

	DBD::TimesTen::st::_prepare($sth, $statement, @attribs)
	    or return undef;

	$sth;
    }

    sub column_info {
	my ($dbh, $catalog, $schema, $table, $column) = @_;

	$catalog = "" if (!$catalog);
	$schema = "" if (!$schema);
	$table = "" if (!$table);
	$column = "" if (!$column);
	# create a "blank" statement handle
	my $sth = DBI::_new_sth($dbh, { 'Statement' => "SQLColumns" });

	DBD::TimesTen::db::_column_info($dbh,$sth, $catalog, $schema, $table, $column)
	    or return undef;

	$sth;
    }
    
    sub table_info {
 	my($dbh, $catalog, $schema, $table, $type) = @_;

	if ($#_ == 1) {
	   my $attrs = $_[1];
	   $catalog = $attrs->{TABLE_CAT};
	   $schema = $attrs->{TABLE_SCHEM};
	   $table = $attrs->{TABLE_NAME};
	   $type = $attrs->{TABLE_TYPE};
 	}

	$catalog = "" if (!$catalog);
	$schema = "" if (!$schema);
	$table = "" if (!$table);
	$type = "" if (!$type);

	# create a "blank" statement handle
	my $sth = DBI::_new_sth($dbh, { 'Statement' => "SQLTables" });

	DBD::TimesTen::db::_table_info($dbh,$sth, $catalog, $schema, $table, $type)
	      or return undef;
	$sth;
    }

    sub primary_key_info {
       my ($dbh, $catalog, $schema, $table ) = @_;
 
       # create a "blank" statement handle
       my $sth = DBI::_new_sth($dbh, { 'Statement' => "SQLPrimaryKeys" });
 
       $catalog = "" if (!$catalog);
       $schema = "" if (!$schema);
       $table = "" if (!$table);
       DBD::TimesTen::db::_primary_key_info($dbh, $sth, $catalog, $schema, $table)
	     or return undef;
       $sth;
    }

    sub foreign_key_info {
       my ($dbh, $pkcatalog, $pkschema, $pktable, $fkcatalog, $fkschema, $fktable ) = @_;
 
       # create a "blank" statement handle
       my $sth = DBI::_new_sth($dbh, { 'Statement' => "SQLForeignKeys" });
 
       $pkcatalog = "" if (!$pkcatalog);
       $pkschema = "" if (!$pkschema);
       $pktable = "" if (!$pktable);
       $fkcatalog = "" if (!$fkcatalog);
       $fkschema = "" if (!$fkschema);
       $fktable = "" if (!$fktable);
       DBD::TimesTen::db::_foreign_key_info($dbh, $sth, $pkcatalog, $pkschema, $pktable, $fkcatalog, $fkschema, $fktable) or return undef;
       $sth;
    }

    sub ping {
	my $dbh = shift;
	my $state = undef;

 	my ($catalog, $schema, $table, $type);

	$catalog = "";
	$schema = "";
	$table = "NOXXTABLE";
	$type = "";

	# create a "blank" statement handle
	my $sth = DBI::_new_sth($dbh, { 'Statement' => "SQLTables_PING" });

	DBD::TimesTen::db::_table_info($dbh, $sth, $catalog, $schema, $table, $type)
	      or return 0;
	$sth->finish;
	return 1;

    }

    # New support for the next DBI which will have a get_info command.
    sub get_info {
	my ($dbh, $item) = @_;
	return DBD::TimesTen::db::_get_info($dbh, $item);
    }

    # new override of do method provided by Merijn Broeren
    # this optimizes "do" to use SQLExecDirect for simple
    # do statements without parameters.
    sub do {
        my($dbh, $statement, $attr, @params) = @_;
        my $rows = 0;
        if( -1 == $#params )
        {
          # No parameters, use execute immediate
          $rows = _ExecDirect( $dbh, $statement );
          if( 0 == $rows )
          {
            $rows = "0E0";
          }
          elsif( $rows < -1 )
          {
            undef $rows;
          }
        }
        else
        {
          $rows = $dbh->SUPER::do( $statement, $attr, @params );
        }
        return $rows
    }

    sub type_info_all {
	my ($dbh) = @_;
	my $sth = DBI::_new_sth($dbh, { 'Statement' => "SQLGetTypeInfo" });
	DBD::TimesTen::db::_type_info($dbh, $sth, DBI::SQL_ALL_TYPES) or return undef;
	my $info = $sth->fetchall_arrayref;
	unshift @$info, {
	    map { ($sth->{NAME}->[$_] => $_) } 0..$sth->{NUM_OF_FIELDS}-1
	};
	return $info;
    }

}


{   package DBD::TimesTen::st; # ====== STATEMENT ======
    use strict;

    sub cancel {
	my $sth = shift;
	my $tmp = DBD::TimesTen::st::_cancel($sth);
	$tmp;
    }
}

1;
__END__

=head1 NAME

DBD::TimesTen - TimesTen Driver for DBI

=head1 SYNOPSIS

  use DBI;

  $dbh = DBI->connect('dbi:TimesTen:DSN=...', 'user', 'password');

See L<DBI> for more information.

=head1 DESCRIPTION

=head2 Notes:

=over 4

=item B<An Important note about the tests!>

 Please note that some tests may fail or report they are
 unsupported on this platform.
   
=item B<Private DBD::TimesTen Attributes>

=item ttIgnoreNamedPlaceholders

Use this if you have special needs where :new or :name mean something
special and are not just placeholder names.  You I<must> then use ? for
binding parameters.  Example:

	$dbh->{ttIgnoreNamedPlaceholders} = 1;
	$dbh->do("create trigger foo as if :new.x <> :old.x then ... etc");

Without this, DBD::TimesTen will think :new and :old are placeholders for
binding and get confused.
 
=item ttDefaultBindType

This value defaults to 0, which means that DBD::TimesTen will attempt to
query the driver via SQLDescribeParam to determine the correct type.  This
parameter is overridden if you supply a data type with the bind_param() call.

=item ttExecDirect

Force DBD::TimesTen to use SQLExecDirect instead of SQLPrepare() then SQLExecute.
There are drivers that only support SQLExecDirect and the DBD::TimesTen
do() override doesn't allow returning result sets.  Therefore, the
way to do this now is to set the attributed ttExecDirect.

There are currently two ways to get this:
	$dbh->prepare($sql, { ttExecDirect => 1}); 
 and
	$dbh->{ttExecDirect} = 1;

When $dbh->prepare() is called with the attribute "ExecDirect" set to a
non-zero value dbd_st_prepare do NOT call SQLPrepare, but set the sth flag
ttExecDirect to 1.
 
=item ttQueryTimeout

This allows the end user to set a timeout for queries.  You can either set
this via the attributes parameter during the connect call, or you can directly
modify the attribute using $dbh->{ttQueryTimeout} = 30 after you have already
connected.

=item ttIsolationLevel

This allows the end user to set the isolation level.  You may need to
commit if you are changing isolation levels.  You can either set this via
the attributes parameter during the connect call, or you can directly
modify the attribute using $dbh->{ttIsolationLevel} after you have already
connected.  You must export the symbols using as follows:

	use DBI;
	use DBD::TimesTen qw(:sql_isolation_options);

	$dbh = DBI->connect('DBI:TimesTen:DSN=...', undef, undef,
	   { ttIsolationLevel => SQL_TXN_SERIALIZABLE });

The valid isolation levels are:

	SQL_TXN_READ_COMMITTED (default)
	SQL_TXN_SERIALIZABLE

   
=head2 Frequently Asked Questions

Answers to common DBI and DBD::TimesTen questions:

=over 4
 
=item Almost all of my tests for DBD::TimesTen fail.  They complain about
not being able to connect or the DSN is not found.  

Verify that you set DBI_DSN, DBI_USER, and DBI_PASS.

=cut
